tsdb/db_test.go

```diff
@@ -3337,6 +3338,42 @@ func TestChunkQuerier_ShouldNotPanicIfHeadChunkIsTruncatedWhileReadingQueriedChu
 	}
 }
 
+func TestAllowOverlappingBlocks_ShouldPanicIfDisableBlocksOverlappingInConfigFileWhen_Reload(t *testing.T) {
+	db := openTestDB(t, nil, nil)
+
+	defer func() {
+		require.NoError(t, db.Close())
+	}()
+
+	db.opts.AllowOverlappingBlocks = true
+	cfg := &config.DefaultConfig
+	cfg.StorageConfig.TSDB = &config.TSDBConfig{
+		Retention: &config.TSDBRetentionConfig{
+			AllowOverlappingBlocks: false,
+		},
+	}
+
+	require.Error(t, db.ApplyConfig(cfg))
+}
+
+func TestAllowOverlappingBlocks_ShouldNotPanicIfDisableBlocksOverlappingInConfigFileWhen_Reload(t *testing.T) {
+	db := openTestDB(t, nil, nil)
+
+	defer func() {
+		require.NoError(t, db.Close())
+	}()
+
+	db.opts.AllowOverlappingBlocks = true
+	cfg := &config.DefaultConfig
+	cfg.StorageConfig.TSDB = &config.TSDBConfig{
+		Retention: &config.TSDBRetentionConfig{
+			AllowOverlappingBlocks: true,
+		},
+	}
+
+	require.NoError(t, db.ApplyConfig(cfg))
+}
```

Comment:
Let's combine these tests as `TestDBApplyConfig` and check if DB states have been updated properly after applying config (for retention and also allow overlapping blocks). (And it's not a panic :) it is just returning an error)